designs/2022-languages/README.md

```diff
@@ -0,0 +1,612 @@
+* [typescript-eslint](https://typescript-eslint.io)
+
+Instead of forcing others to rewrite what is core linting functionality, or asking people to produce an ESTree format AST for some language other than JavaScript, we can make everyone's lives easier by providing formal language support such that languages can be plugged in to ESLint easily.
+
+We did also previously accept [#56](https://github.com/eslint/rfcs/pull/56) as an extension to `parseForESLint()` to support more languages.
+
+## Detailed Design
+
+This proposal consists of the following changes:
+
+1. Allows plugins to publish languages in addition to rules, processors, etc.
+1. Allows end users to select which language to use in their config files.
+1. Defines an API for defining languages.
+1. Changes the ESLint core to consume language definitions instead of assuming JavaScript is always used.
+
+**Prerequisites:** This proposal requires the flat config system and cannot be used with the eslintrc config system.
+
+**Note:** This design is intended to work both in the current version of ESLint and also in the "rewrite" version, which I'll refer to as ESLint X to distinguish between the two. Not all features of this proposal will be available in both versions. 
+
+### Languages in plugins and configs
+
+Languages can be defined in plugins by using the `languages` key and giving the language a name. This allows multiple languages to be contained in a single plugin if desired. For example:
+
+```js
+// inside a plugin called eslint-plugin-lang
+exports default {
+
+    languages: {
+        lang1: {
+            // language object
+        },
+        lang2: {
+            // language object
+        }
+    },
+    processors: {
+        // ..
+    },
+    rules: {
+        // ...
+    }
+};
+```
+
+Here, `lang1` and `lang2` are languages defined by the plugin. End users can specify which language to use via the `language` key in a config object, such as:
+
+
+```js
+// inside eslint.config.js
+import lang from "eslint-plugin-lang";
+
+exports default [
+    {
+        plugins: {
+            lang
+        },
+        files: ["**/*.lang"],
+        language: "lang/lang1",
+        languageOptions: {
+            // options
+        }
+    }
+];
+```
+
+Here, `"lang/lang1"` refers to the `lang1` language as exported from `eslint-plugin-lang`. Anything inside of `languageOptions` is then handled by `lang1` rather than by ESLint itself. 
+
+### Language Definition Object
+
+Each language definition object must implement the following interface:
+
+```ts
+interface ESLintLanguage {
+
+    /**
+     * Indicates how ESLint should read the file.
+     */
+    fileType: "text" | "binary";
+
+    /**
+     * Validates languageOptions for this language.
+     */
+    validateOptions(options: LanguageOptions): void;
+
+    /**
+     * Parses the given file input into its component parts.
+     */
+    parse(file: File, env: LanguageContext): ParseResult | Promise<ParseResult>;
```

Comment:
I wasn't attempting to tackle multi-file awareness in this proposal, mostly because I still don't have a clear definition of "multi-file awareness." :smile: The C use case seems to be an extreme one, where we can't know the syntax without knowing where the macro definition is.

The intent of this proposal was to allow support for languages primarily used in web application development, where they would likely already have ESLint installed. I'm not sure C would fall into that category.

If you have an idea of what multi-file awareness means and how you could envision it working in ESLint, I'm definitely open to hearing it. If not, then I'd suggest we not worry about it and instead focus on what folks would need to for languages like HTML, CSS, JavaScript, JSON, YAML, Markdown, etc.

Reply:
I'm okay saying multi-file languages are out of scope for this particular RFC.

At a _very_ high level, I imagine multi-file languages would have an API that allows ESLint to ask what a file's dependencies are to build up the dependency graph upfront. Linting would then proceed in dependency order, and the language would need to be able to hand ESLint some additional data (e.g. a description of exports) for each file after linting that ESLint would make available to dependents.

Reply:
Ah, interesting. I see a bit of a chicken-and-egg issue: we would need to parse the file to discover the imports, but if parsing requires the imports, then we are a bit stuck. At least with JavaScript, the syntax is known ahead of time so we can parse the whole program, pull the import info from the AST, and then pull in those imports.

I suppose we could add a prescan step for imports...but yeah, it seems like maybe this should be a followup RFC instead of trying to figure out how to do it here?

Reply:
See https://github.com/eslint/rfcs/pull/105 😄

Reply:
@JamesHenry Would you mind opening a discussion about this use case? I'd like to understand more about this relationship and what you'd like to accomplish.